### ISSUE TITLE
ADT - How can I align pointcloud in scene coordinate?

### ISSUE THREAD

**Issue by choiszt** [facebookresearch/projectaria_tools#27]
Hi team!
Currently I'm reconstructing the 3D pointcloud scene based on your brilliant guidance in adt_quickstart_tutorial.ipynb. However, it comes to some issues about converting matrices when I was implementing the pointcloud conversion.

Firstly, I supposed that the depth data are in the camera coordinate based on 
`gt_provider.get_depth_image_by_timestamp_ns(timestamp, stream_id)`
and then I acquire the intrinsic data from 
```
calibration_provider=gt_provider.get_aria_camera_calibration(stream_id)
fx,fy=calibration_provider.get_focal_lengths()
cx,cy=calibration_provider.get_principal_point()
```
in order to convert rgbd data to pointcloud as follows:

```
def rgbd_to_ply(rgb_path, depth_path, ply_path,fx,fy,cx,cy):
    rgb = cv2.imread(rgb_path)
    depth = cv2.imread(depth_path, cv2.IMREAD_UNCHANGED)

    if rgb is None or depth is None:
        raise ValueError("Load Error")

    h, w= depth.shape

    fx = fx 
    fy = fy
    cx = cx
    cy = cy

    vertices = []
    for y in tqdm(range(h)):
        for x in range(w):
            Z = depth[y, x] / 1000.0 
            X = (x - cx) * Z / fx
            Y = (y - cy) * Z / fy
            color = rgb[y, x]
            vertices.append((X, Y, Z, color[2], color[1], color[0]))  # BGR->RGB

    vertex = np.array(vertices, dtype=[('x', 'f4'), ('y', 'f4'),('z', 'f4'), ('red', 'u1'),('green', 'u1'), ('blue', 'u1')])
    el = PlyElement.describe(vertex, 'vertex')
    PlyData([el]).write(ply_path)
```

The next step I want to align pointclouds in each frame to the same world coorindate, so I was trying to acquire the extrinsic matrices:
```
cam2dev=gt_provider.get_aria_transform_device_camera(stream_id).inverse().to_matrix() #camera 2 device
dev2sce=gt_provider.get_aria_3d_pose_by_timestamp_ns(STAMP).data().transform_scene_device.inverse().to_matrix()#device 2 scene
```
and I supposed `transform_matrix=cam2dev@dev2sce `is the exactly extrinsic matrix.

Finally, I implement the alignment in world coordinate by 
`def transform_ply_with_matrix(ply_path, transform_matrix, output_path):
    ply_data = PlyData.read(ply_path)
    vertex_data = np.vstack([ply_data['vertex']['x'], ply_data['vertex']['y'], ply_data['vertex']['z']]).T

    homogenous_coords = np.hstack([vertex_data, np.ones((vertex_data.shape[0], 1))])

    transformed_data = (transform_matrix @ homogenous_coords.T).T

    ply_data['vertex']['x'] = transformed_data[:, 0]
    ply_data['vertex']['y'] = transformed_data[:, 1]
    ply_data['vertex']['z'] = transformed_data[:, 2]

    ply_data.write(output_path)`
but the mismatch encountered when I tried to visualize two ply in Meshlab.

**Comment by nickcharron** [comment]
Hi @choiszt,

I think you may be misinterpretting our transformation notation here:

```
cam2dev=gt_provider.get_aria_transform_device_camera(stream_id).inverse().to_matrix() 
dev2sce=gt_provider.get_aria_3d_pose_by_timestamp_ns(STAMP).data().transform_scene_device.inverse().to_matrix()
```

Please see our documentation on frame convention [here](https://facebookresearch.github.io/projectaria_tools/docs/data_formats/coordinate_convention/3d_coordinate_frame_convention#a-note-on-sensor-naming-and-motivation)

tl;dr:

gt_provider.get_aria_transform_device_camera(stream_id) returns the transformation from camera to device. I believe you have the inverse in your example code.

**Comment by choiszt** [comment]
Hello @nickcharron,
really appreciate for your response!
So I guess the correct transformations are:
```
T_Dev_Cam=gt_provider.get_aria_transform_device_camera(stream_id).to_matrix()  #T_Device_Camera
T_Sce_Dev=gt_provider.get_aria_3d_pose_by_timestamp_ns(STAMP).data().transform_scene_device.to_matrix() #T_scene_device
T_Sce_Cam=T_Sce_Dev@T_Dev_Cam #transform from Cam to Sce
pcd_in_Scene  = (T_Sce_Cam@ homogenous_coords.T).T
```

But still have mismatch in Meshlab
![image](https://github.com/facebookresearch/projectaria_tools/assets/76559926/ed9daaac-795d-4030-921e-72e4f39ebf11)

**Comment by choiszt** [comment]
Hi @nickcharron !
I'm not sure if I've made myself clear. The unproject function seems to project an x,y tensor(with the shape [2,1]) to the 3D coordinate. I suppose projecting each pixel in images to the 3D coordinate as a time-consuming operation.
I do have some inspiration from the "undistort" concept, and I implement as follows:
```
def undistort(image):
    sensor_name = gt_provider.raw_data_provider_ptr().get_label_from_stream_id(stream_id)
    device_calib = gt_provider.raw_data_provider_ptr().get_device_calibration()
    src_calib = device_calib.get_camera_calib(sensor_name)
    dst_calib = calibration.get_linear_camera_calibration(1408, 1408, calibration_provider.get_focal_lengths()[0], sensor_name)
    # rectify image
    rectified_image = calibration.distort_by_calibration(image, dst_calib, src_calib)
    return rectified_image
```
And I suppose the intrinsic matrix are compatible to the undistort RGBD.
But unfortunately, the mismatching still occur. BTW the point of pinhole radius occurred, I guess **distort_by_calibration** may have something related to **unproject function**
![image](https://github.com/facebookresearch/projectaria_tools/assets/76559926/5b3b21f9-2274-40f5-89a9-31c71bd0ff30)

**Comment by chpeng-fb** [comment]
Hi @choiszt, thank you for asking. What are your intrinsics specifically? Since you undistorted the image to a pinhole camera model with image width = 1408, height = 1408 and focal = calibration_provider.get_focal_lengths()[0]. Your fx, fy should equal calibration_provider.get_focal_lengths()[0] and cx,cy = 1408 / 2 = 704. 

The function `distort_by_calibration` just unprojects all the pixels to ray and reproject back with the dst_calib, which in your case is the pinhole camera model. This function is speeded up in the c++ implementation but you can also try the unproject function in the src_calib without undistortion to debug where the issue is. 

Plus, if you could share your complete code, it might be easier to check where could be the error.

**Comment by nickcharron** [comment]
@choiszt I also created an [example notebook](https://colab.research.google.com/drive/1rG9DbPPwK3kP9dgwSR7kUXvd0KUDnjE5?usp=sharing) that you can follow.

One thing to make sure (a mistake I made myself when creating this example) is not to use the get_visualizable() function on the depth images, as that changes the units. 

Hope this helps!

**Comment by choiszt** [comment]
> @choiszt I also created an [example notebook](https://colab.research.google.com/drive/1rG9DbPPwK3kP9dgwSR7kUXvd0KUDnjE5?usp=sharing) that you can follow.
> 
> One thing to make sure (a mistake I made myself when creating this example) is not to use the get_visualizable() function on the depth images, as that changes the units.
> 
> Hope this helps!

Hi @nickcharron !
I've also noticed the get_visualizable() issue when I tried to convert and save data, I'll check out your examples and really appreciate your effortless help!

**Comment by nickcharron** [comment]
Closing this issue, but please feel free to re-open if you have any more questions. Thanks!

**Comment by VitorGuizilini-TRI** [comment]
Hi, quick related question here. From what I understand, undistorting a depth map with `distort_by_calibration` will only modify the location of depth pixels, not the depth values themselves. That's why after undistortion we still cannot produce a pointcloud from the rectified depth map using pinhole intrinsics. Is there a way to return the updated depth values during the undistortion process?  Or produce those some other way? 

Thank you very much!

**Comment by nickcharron** [comment]
Hi @VitorGuizilini-TRI,

Undistorting a depth image is similar to undistoring an RGB image. Using the undistorted image with the pinhole model will result in back projecting the same pixel values to the same physical 3d coordinate as doing the same with the original image + original calibration. So yes, you should be able to produce a pointcloud using the undistorted image with a pinhole model.

The undistortion process does require some slight editing of the pixel values because it needs to fill empty pixels. For those pixels, we provide the option to use a bilinear interpolation or pick the nearest neighbor for getting the depth of that pixel. This will introduce some error. The only way to completely remove that error would be for us to re-render the scene and create new depth maps using a pinhole camera model which would be a significant amount of work.

**Comment by JiajieLi7012** [comment]
> @choiszt I also created an [example notebook](https://colab.research.google.com/drive/1rG9DbPPwK3kP9dgwSR7kUXvd0KUDnjE5?usp=sharing) that you can follow.
> 
> One thing to make sure (a mistake I made myself when creating this example) is not to use the get_visualizable() function on the depth images, as that changes the units.
> 
> Hope this helps!

Hello, I'm trying to first undistort the image and depth map and then create a point cloud using the rectified image and depth map. I'd like to confirm the main modification required in the demo notebook is to use `ray = dst_calib.unproject([u,v])` given that `dst_calib = calibration.get_linear_camera_calibration(512, 512, 280, sensor_name)`. Thanks in advance.

**Comment by nickcharron** [comment]
@JiajieLi7012 correct, that would give you the undistorted calibration, but you would also have to undistort the image and use those u,v values. 

Why do you want to undistort the images when building a pointcloud? Using an undistorted image with the rectified camera model to generate a 3D pointcloud as you're showing will essentially give you the same results as the demo. The only difference is that your approach will introduce artifacts because undistorting images is never perfect and requires interpolation.

**Comment by JiajieLi7012** [comment]
@nickcharron Hi, thanks for your quick and detailed response. As I needed to establish connections/mappings from certain pixels (based on a mask) in the rectified images to the points in the 3D point cloud, I thought it might be more straightforward to directly get a point cloud from rectified images. 

But as you mentioned, the point cloud from rectified images will always have some more artifacts, now I'm thinking about first establishing the mappings from the pixels in the distorted (original) images to the point cloud generated from them and then rectifying these pixels into the undistorted images. Do you think this idea works?
